### ISSUE TITLE
ENH:  news feed

### ISSUE THREAD

**Issue by jpcsupplies** [jpcsupplies/Economy_mod#70]
Was just pondering what I liked about other 4x games trade, and one thing that stuck in my mind was "News reports"

In other single player games, events that occur in game that the player was involved in;   get written to a game "news paper"  where it often lists the players name and participation.

The usefulness of this is probably restricted by what we can and cannot detect in game from modAPI.

In a simple sense this could just be an admin triggered message, that pops up in game; and is written by an admin (for notifying players about events etc)
It should probably check if a player is in a ship or not; and piloting or passenger, so pop up boxes dont open when a player is probably trying not to crash into something.  Just message the screen in that case or wait until they get out next.

It would work similarly to a MOTD, but can be triggered at any time the admin wants, not just on connect.

The news could also be semi-automated, for example when Economy detects certain events occuring, it logs them into a "news" log which players can view any time (or subscribe to for a fee)  with /news

This could also be a useful application of player contracts for example -
"Newsflash:  The infamous criminal  Bob the Griefer was caught and killed in a recent skirmish with Screaming Angel.  For their valor; they were awarded a 5000 {credits} bounty"

** {credits would be whatever name the currency was given in the config file}
*** {probably need to use range check to see who was nearest when player was killed}

This could also use range checks for hostile activity reports -
Another example:
 "Pirates attack!"    Pirate activity has been detected in the area of {trade region name}   {reportedly Screaming Angel fought them off; neutralizing {2} ships..}

Another example:
Declare war event:
Galactic war!:   Faction {x} just declared war on {y};  {Player}, the leader of {x} was quoted as saying - we cannot tolerate {y} any more!

Alliance request event:
Peace talks!  Faction {x} has made overtures to faction {y} to form an alliance..  time will tell where these negotiations lead.

Alliance accept event:
Trade alliance!  Faction {x} just formed an alliance with {y}.   Both leaders were quoted as hoping this new alliance will be mutually beneficial to our trade.

Alliance removes at war flag
Galactic Peace!  Faction {x} is no longer at war with {y}!

Player with most kills?
Top Gun!   Screaming Angels has clocked up {4} kills!

When a cockpit is hacked and a ship stolen
Theft! {faction/player} {cruiser/ship} (large/small)  has been reported stolen! 

Etc..

These might not all be practical, but it should be a good enhancement for role play heavy servers.

**Comment by jpcsupplies** [comment]
Probably should have time stamps too

**Comment by jpcsupplies** [comment]
I just thought of a feature of this that is so obvious I am kicking myself..

When a player enters a new trade zone - i was thinking have the game generate the players a GPS point for it..  (easy done just like your scanore command)      but also pop up a greeting message like the trade station just hailed your etc...
This message could pop up on first encountering the trade zone when it sets the GPS point and display a brief greeting the owner of the zone can configure.     On subsequent visits, just display the message in chat.

This also means we probably should do something similar on the NPC trade zone, although that would be obviously set by the server owner.  Thoughts?  @midspace ?

**Comment by midspace** [comment]
Yes, I was thinking of a "Welcome to the xxxx trade zone" message.

That's why the `MarketStruct` has a `DisplayName`.
It must be defined when the Market is created.
https://github.com/jpcsupplies/Economy_mod/blob/master/Economy/Data/Scripts/Economy.scripts/EconStructures/MarketStruct.cs

My only concern was how often we do a check to determine when a player has entered a zone.
It can't be too often, as it'll be cycling each player for each zone.
Unless, we palm this processing off to the client side, but then we'll have to start passing zone location data to the client when a zone is created or deleted, or moved.
When you are talking about a moving trade zone, it starts to get costly to transfer that detail to clients every time it moves.

And then, you might have a player that doesn't want to send a notification message, because they operate like a black market.

And you could have a "You are now leaving the xxxx trade zone. Have a nice day!" message.

**Comment by jpcsupplies** [comment]
Ok so -
1: a /global command that a: sends a mission screen message to all connected players, and b: writes this message to a /news log. 
2: a notification when a player enters a trade region
3: a /news command that displays the last 10 or so global messages with a time stamp
4: when a player creates a new trade zone - this event should go into news log too

**Comment by jpcsupplies** [comment]
Beacon idea sounds pretty good actually, palms of the process from our script entirely.. if a player is in a trade region bam there it is.  Its how the beacons are /manually/ set up on Ramblers frontier in the trade zone now.   Only downside there is some players discussed the possibility of a station or crate being related to their trade zone - a beacon makes it pretty obvious where it is.  But if we dont use crates it wont matter.

The gps idea has merit too, as the players can look it up and even set autopilot or jump points with it when they need to travel there.    Is there any easy way to add a gps point to a player?   I assume so - that is how your metal detector.scanner worked.

**Comment by midspace** [comment]
http://forums.keenswh.com/threads/modding-api-new-requests.7141893/page-7#post-1286942455
> Then you can use something like this to play a sound:

```
private MyEntity3DSoundEmitter m_soundEmitter;
		public MyEntity3DSoundEmitter SoundEmitter { get { return m_soundEmitter; } }
		m_soundEmitter = new MyEntity3DSoundEmitter(Container.Entity as VRage.Game.Entity.MyEntity);
 
		public static void PlaySound(this IMyFunctionalBlock ftl, string soundname, bool stopPrevious = false)
		{
			MyEntity3DSoundEmitter emitter = null;
 
			if( ftl.IsFTL() )
				emitter = ftl.GameLogic.GetAs<FTL>().SoundEmitter;
			if( emitter != null )
			{
				if (string.IsNullOrEmpty(soundname))
				{
					emitter.StopSound((emitter.Loop ? true : false), true);
				}
				else
				{
					MySoundPair sound = new MySoundPair(soundname);
					emitter.CustomMaxDistance = (float)ftl.GetTopMostParent().PositionComp.WorldVolume.Radius * 2.0f;
					Logger.Instance.LogDebug("Distance: " + emitter.CustomMaxDistance);
					emitter.CustomVolume = 1.0f;
					emitter.PlaySound(sound, stopPrevious);
				}
			}
		}
```

**Comment by jpcsupplies** [comment]
Ah. so we call playsound ((from a block?), name of sound goes here,) 
we get 3dsoundemitter   create an emitter point, pair it with a sound..   i think it grabs the volume setting.. logs the sound playing then triggers the sound, stopping any that were already playing..

Assuming we can play a sound added to the workshop mod, that should do it with a bit of fiddling..

i gather it makes more sense to you as you are more familiar with these various object references.

This example is pairing it to an FTL block and letting the game do all the 3d positional sound business...      so i assume we could adapt it to attach it to one of our [economy] tagged LCDs which would work more or less as I was hoping..      means anyone on the ship hears it too ?

**Comment by jpcsupplies** [comment]
Ok here is where we stand here:
Stage 1 news system: /global (admin) and /news (players)
- /global pops up on everyones screen  (may want some logic to prevent piloted ships getting it in their face.. maybe just a text scroller for them instead of the pop up)
- Globals are written with a time stamp to the /news log. Which holds the last 50 or so (configurable) events?
- Some sort of visual and audio indication to a player when they enter a trade zone. Optionally auto create a GPS point 
- An entry added to the news log when a new trade zone is created.  NPC zones give GPS location. Players zones just list the player/faction name, no GPS with some roleplay sounding phrasing.  "X (player) of Y (Faction) has just founded a new trade company X km/m (distance) from Z (trade network name - referencing 0,0,0 NPC trade zone)"   and...      "Z (trade network name) launches a new trade portal at (name of NPC zone) GPS Coords"

Stage 2 news system
- Add a way for a /global to optionally generate a GPS point for all online players.  (the greyed out style ones you get from chats or lcds)  OR...
 - ...fine tune the news system to specifically allow role play style news events (which always generate gps points) to be posted seperate to normal globals- (eg in the case of contracts/subsidies, pirate warnings etc)
- the role play events can be created manually by admins, or triggered by some sort of automatical mission/contract/subsidy system?  (eg.  Station X at GPS desperately requires resource Z, the first player supplying this resourse will receive a bonus x% on top of the normal price for Y (7?) days)  probably need an accociated /contracts command for players to list the details of their currently held contracts (and the hud contract count read out)
- Add a way that when /news is used any news entries with associated GPS points will re-create these "temp" greyed GPS points to.
- Economic news?  Eg when a player is sent a payment request and they cannot afford it before the time out, it creates a "x player is having financial difficlties"?  or when a market runs out of stock or money? or X player just sucessfully sold a ship|station to Y player
War news?  Can we catch when players or turrents lock on or fire at other players?
Abandoned! Purge logic: can we create an entry indicating "X (player) has abandoned the area/sector..  (network name) seizes his funds and supplies, derelicts may exist!"   when a given player exceeds the account purge timeout.
- Exploration news -  the first player to enter a given trade zone ever gets reported as "XX trade zone celebrates the arrival of its first customer"  (sound familiar? *chuckles* )  or similar first buy/sell/delivery type news

**Comment by jpcsupplies** [comment]
Probably also need an on-connect and in-game  notification along the lines of a message stating:
You have X unread news items.
You have received a new news item.

Etc
